### PR TITLE
Second databases chapter.

### PR DESCRIPTION
**Pull request by jimflores5**

### CODE REVIEW

File: src/assignments/flask-game.rst [hunk 1]
```diff
@@ -0,0 +1,6 @@
+.. _flask-game:
+
+Assignment #6: Flask Game
+=========================
+
+Lorem ipsum...
```
Comment (codinglikeagirl42) [inline src/assignments/flask-game.rst:6]: Assignment #6 was already merged, will this cause a problem?

File: src/chapters/databases-part2/introduction.rst [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+   <section class="vid_box">
+      <iframe class="vid" src="https://www.youtube.com/embed/n2tXf33yAQk" frameborder="1" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
+   </section>
+
+Project Setup
+-------------
+
+#. Launch Visual Studio Code and open your ``local_practice`` directory.
+#. Clone the `Minesweeper starter code <https://github.com/LaunchCodeEducation/LCHS_minesweeper>`__
+   from GitHub. The project folder is called ``LCHS_minesweeper``.
```
Comment (codinglikeagirl42) [inline src/chapters/databases-part2/introduction.rst:35]: So I originally downloaded the repo to my computer and had to create the gitignore file, not sure why?

I just tested cloning and it is there.  Not sure if we need to make a note or not.

File: src/chapters/databases-part2/db-setup.rst [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+For starters, our database should save:
+
+#. Information about each mine hidden on the board. This includes a primary key
+   value as well as a location (row letter plus column number).
+#. Information about each individual cell on the board. This includes a primary
+   key, the cell's location, the number of mines in surrounding cells, whether
+   the user has clicked the cell, and whether the cell contains a mine.
+
+This outline suggests at 2 tables in the database - one for a list of mines,
```
Comment (codinglikeagirl42) [inline src/chapters/databases-part2/db-setup.rst:29]: Weird wording - This outline suggests at 2 tables

File: src/chapters/databases-part2/db-setup.rst [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+and another for the cells on the game board.
+
+``db_setup.py``
+---------------
+
+Use the tools in Visual Studio code to create a new Python file at the same
+level as ``main.py``. Call the new file ``db_setup.py``. This will be a
+separate program, and its only job is to create the database file and add two
```
Comment (codinglikeagirl42) [inline src/chapters/databases-part2/db-setup.rst:37]: The db_setup.py already exists when cloned

File: src/chapters/databases-part2/home-page.rst [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+Note the following:
+
+#. We link to a ``style.css`` file in line 8. It contains some basic styling
+   rules for the game, and the tutorials won't instruct you to change any of
+   the settings. However, you absolutely CAN adjust the properties and values
+   as you see fit! Don't like the gray background on the board? Fix it!
+#. On line 13, the placeholder for the page title needs to be filled. Don't
+   forget to pass a value to ``page_title`` from each ``render_template()``
+   function.
```
Comment (codinglikeagirl42) [inline src/chapters/databases-part2/home-page.rst:46]: A link to instructions for review would be nice

File: src/chapters/databases-part2/home-page.rst [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+#. Feel free to add extra elements to ``base.html``. Remember that this
+   content will appear on any template that extends from the base.
+
+Build ``index.html``
+--------------------
+
```
Comment (codinglikeagirl42) [inline src/chapters/databases-part2/home-page.rst:52]: For the life of me I could not get it to display 5-25 to select a number.  Finally changed min width and played with the number to get it to work.

File: src/chapters/databases-part2/home-page.rst [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+   <section class="vid_box">
+      <iframe class="vid" src="https://www.youtube-nocookie.com/embed/iKkYGMHWWco" frameborder="1" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
+   </section>
+
+Video Summary
+-------------
+
+#. Our job right now is to build the view for the home page. We won't worry
+   about adding any code to deal with placing mines or collecting data from
+   the form.
```
Comment (codinglikeagirl42) [inline src/chapters/databases-part2/home-page.rst:79]: Found this redundant form the lines above 62 and 63,  not sure if I'd reword or just delete?

File: src/chapters/databases-part2/game-page.rst [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+#. Inside the same form, add a *Flag Mine* checkbox above the game board, and
+   center it.
+#. In the ``<header></header>`` element, add the *New Game* button.
+
+   a. Style the ``<button>`` with the ``menu`` class.
+   b. Give it the same ``name`` and ``type`` as the buttons in the table.
+   c. Set the ``value`` to be the empty string.
+   d. Use the ``form`` attribute to match the button to the
+      ``<form id="game-board" ...>`` element.
+
```
Comment (codinglikeagirl42) [inline src/chapters/databases-part2/game-page.rst:64]: I really wanted to add the link to the main page here as it makes sense in my mind.  Im guessing your logic was to code all the buttons at the same time, 25.7.1.1?  I'm just use to setting up my web pages and linking everything first and then adding the content to the pages.

File: src/chapters/databases-part2/game-page.rst [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+   c. Set ``type="submit"``,
+   d. Use a placeholder to assign a ``value`` that matches the cell coordinates
+      (A10, B4, C7, etc.).
+
+#. Inside the same form, add a *Flag Mine* checkbox above the game board, and
+   center it.
+#. In the ``<header></header>`` element, add the *New Game* button.
+
+   a. Style the ``<button>`` with the ``menu`` class.
+   b. Give it the same ``name`` and ``type`` as the buttons in the table.
```
Comment (codinglikeagirl42) [inline src/chapters/databases-part2/game-page.rst:60]: I think I understand your logic for giving it the same name and type.  To me this button just needed to take us to the home page.  So could we have made it separate? Also I did play around with using different strings besides empty, as long as it wasn't a board string just to see if they would work.  They did!

File: src/chapters/databases-part2/home-logic.rst [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+To complete your ``index()`` function, open your Minesweeper project and code
+along with the video:
+
+.. raw:: html
+
+   <section class="vid_box">
+      <iframe class="vid" src="https://www.youtube-nocookie.com/embed/OZ1-PDXDZwI" frameborder="1" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
+   </section>
+
```
Comment (codinglikeagirl42) [inline src/chapters/databases-part2/home-logic.rst:26]: at 1.22min in the video:

def record_mines(locations) = def record_mines(coords) in starter code

File: src/chapters/databases-part2/game-logic.rst [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+   the board before hiding any mines. To prevent this, we redirect them back to
+   the home page if they arrive at the page through a ``GET`` request.
+#. We need to collect two pieces of information from the game board: which cell
+   the player selected and whether or not they clicked the *Flag Mine*
+   checkbox.
+
+   a. Each button on the board has the same name, and the value for each one
+      matches its cell coordinates (like ``A10``).
+   b. However, the value for the *New Game* button is the empty string.
+   c. The checkbox input has a different name from the buttons.
```
Comment (codinglikeagirl42) [inline src/chapters/databases-part2/game-logic.rst:49]: I enjoyed playing around assigning different strings

File: src/assignments/flask-game.rst (hunk 1 above)
Comment (jimflores5) [reply src/assignments/flask-game.rst:6]: Nope!  We're good.

File: src/chapters/databases-part2/home-logic.rst [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+#. Use a conditional to break the ``index()`` function into two parts. One part
+   deals with the form submission (a ``POST`` request), and the other resets
+   the board after a ``GET`` request.
+#. After a ``POST`` request, use ``request.form[]`` to collect data from the
+   *Main Menu* form. This will be the number of mines the user wants to hide.
+
+   a. Add some server-side validation to deal with invalid form submissions
+      (e.g. submitting and empty field, or letters, or symbols, etc.).
+   b. For invalid form submissions, set the number of mines to some default
+      value, like ``10``.
```
Comment (codinglikeagirl42) [inline src/chapters/databases-part2/home-logic.rst:47]: when you change tie type to text, 3 and 99 get thru, doesn't break the game but might be a missed validation

File: src/chapters/databases-part2/introduction.rst (hunk 2 above)
Comment (jimflores5) [reply src/chapters/databases-part2/introduction.rst:35]: I just double checked the clone sequence, and the `.gitignore` file came with the download.  I think we can leave this one alone.

Even if the file doesn't clone to the kids computers, they will still be able to run the program.  Of course, their Flask, environment, and database files will all git recorded by the VCS.  At this point in the course, I think we can leave it to them to correct any glitches.

File: src/chapters/databases-part2/db-setup.rst (hunk 3 above)
Comment (jimflores5) [reply src/chapters/databases-part2/db-setup.rst:29]: Fixed!

File: src/chapters/databases-part2/game-logic.rst [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+yet. In this chapter, we focused on getting the game up and running, and we
+accomplished that goal. However, the project does contain a number of
+shortcomings. How to fix these is up to you! You have a good starting place
+right now. Your ideas will determine the next step.
+
+.. admonition:: Note
+   
+   No project is ever really done. Unusual bugs will pop up that need to be
+   fixed, or new feature ideas will pop into your brain. Coding an application
+   is a continuous process of tinkering, testing, and discovery!
```
Comment (codinglikeagirl42) [inline src/chapters/databases-part2/game-logic.rst:161]: Not sure if you would want to mention minimum viable product (MVP), it is something I we stressed a lot when teaching game design and apps. Maybe rework lines 151 to 155 just to introduce the term.  Students who are interested will find their way to google.

This was one of the videos we used https://www.youtube.com/watch?v=UvCri1tqIxQ.  Might be too much but I always thought exposing the students to industry thinking was good.  We talked a lot about scope and you've done a great job of showing how to create a game step by step!

File: src/chapters/databases-part2/db-setup.rst [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+   increment the primary key values).
+
+   ``coordinates`` is of type ``TEXT`` and will store values like ``C9`` or
+   ``E2``.
+#. The ``board`` table includes columns for ``cell_id`` (the primary key),
+   ``coordinates``, ``surr_mines`` (surrounding mines), ``guessed`` and
+   ``mine_id``.
+
+   ``cell_id``, ``surr_mines``, and ``mine_id`` are integers. ``coordinates``
+   is a text data type, and ``guessed`` is a boolean.
```
[the remark this answers is not in the document]
Comment (jimflores5) [reply src/chapters/databases-part2/db-setup.rst:91]: Crud.  I should have used `BOOL` for `guessed` in the `sql_query` string.

`sqlite3` doesn't actually have a boolean data type, so T/F automatically get converted to the integers 0 and 1.  That's what happened here, although TEXT (maybe) instead of INT.

I'm going to see if I can overlay the word BOOL over TEXT in the clip.  That might save me from having to redo the whole thing!

File: src/chapters/databases-part2/home-page.rst (hunk 5 above)
Comment (jimflores5) [reply src/chapters/databases-part2/home-page.rst:46]: Done!

File: src/chapters/databases-part2/home-page.rst (hunk 6 above)
Comment (jimflores5) [reply src/chapters/databases-part2/home-page.rst:52]: I'm not sure what you mean by not getting 5 - 25 to display.  Are you talking about the grey placeholder text not showing up, the size of the input field, the range of allowed numbers, etc.?

File: src/chapters/databases-part2/home-page.rst (hunk 7 above)
Comment (jimflores5) [reply src/chapters/databases-part2/home-page.rst:79]: I reworded the bullet point.  Let me know how it reads.

File: src/chapters/databases-part2/game-page.rst (hunk 9 above)
Comment (jimflores5) [reply src/chapters/databases-part2/game-page.rst:60]: Using a button instead of a link was a choice I made to accommodate my older eyeballs.  I can see and click a button much easier than a link.  Also, attaching the button to the form element means not messing around with explaining how to add an `href` to a button, or how to make an `<a></a>` element look like a button.  So, personal preference.  ;-)

You're right that any string other than a board coordinate will work for the New Game button.  I just defaulted to the empty string (which in a pinch translates to `False` in a conditional).

File: src/chapters/databases-part2/home-logic.rst (hunk 10 above)
Comment (jimflores5) [reply src/chapters/databases-part2/home-logic.rst:26]: Good catch!  I fixed the starter code to match.

File: src/chapters/databases-part2/game-logic.rst (hunk 11 above)
Comment (jimflores5) [reply src/chapters/databases-part2/game-logic.rst:49]: Tinkering is always fun!

File: src/chapters/databases-part2/game-logic.rst (hunk 13 above)
Comment (jimflores5) [reply src/chapters/databases-part2/game-logic.rst:161]: MVP was included in an earlier draft of the chapter (in the introduction, as it turns out), but I ultimately cut the definition and references.  As you said, it might be a bit too much.  There are so many details flying around in this chapter, that I wanted to avoid saturation.

However, your comment did allow me to see that **minimum viable product** was still in the index and led to a blank spot on the first page, so I removed it!  :-)

File: src/chapters/databases-part2/home-logic.rst (hunk 12 above)
Comment (jimflores5) [reply src/chapters/databases-part2/home-logic.rst:47]: Doh!  Busted!

Adding more server-side validation is an option listed on the Next Steps page, but I'll add a note here to point out the shortcoming.

File: src/chapters/databases-part2/db-setup.rst (hunk 4 above)
Comment (jimflores5) [reply src/chapters/databases-part2/db-setup.rst:37]: I removed `db_setup.py` from the starter code since the video walks through its development from start to finish.

File: src/chapters/databases-part2/home-page.rst (hunk 6 above)
Comment (codinglikeagirl42) [reply src/chapters/databases-part2/home-page.rst:52]: Yes, my placeholder text display only showed the 5 because the box was too small.

File: src/chapters/databases-part2/db-setup.rst [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+   </section>
+
+.. admonition:: Note
+
+   For our basic game, we might not use all of the columns in the tables yet.
+   However, they could be useful later, so there's no harm in adding them now.
+
+Video Summary
+-------------
+
```
Comment (codinglikeagirl42) [inline src/chapters/databases-part2/db-setup.rst:63]: I started from scratch - just a note, typing in python db_setup.py throws an error for me, I forgot to try python3 db_setup.py, I just hit the run button.  I do think that at this point students should be able to actively debug but just wanted to make a note.

Comment (jimflores5) [reply src/chapters/databases-part2/db-setup.rst:63]: Corrected in revised video.

File: src/chapters/databases-part2/home-logic.rst [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+along with the video:
+
+.. raw:: html
+
+   <section class="vid_box">
+      <iframe class="vid" src="https://www.youtube-nocookie.com/embed/OZ1-PDXDZwI" frameborder="1" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
+   </section>
+
+Video Summary
+-------------
```
Comment (codinglikeagirl42) [inline src/chapters/databases-part2/home-logic.rst:28]: Video 4 - 1:28 min to 4:29min you are on the game_logic.py file and guessed = null 
at 2:00 you mention all set to Null, but you mention that its starting a clean slate, I don't see an issue with False as the value here if you do the black bar thingie.(no scrolling)

9:43 you go back to game_logic.py but scroll down pretty fast

File: src/chapters/databases-part2/db-setup.rst (hunk 14 above)
[the remark this answers is not in the document]
Comment (jimflores5) [reply src/chapters/databases-part2/db-setup.rst:91]: I wound up re-recording the video.  It's 1:10 min longer than the original, but it properly sets `guessed` as a boolean data type now.

File: src/chapters/databases-part2/home-logic.rst (hunk 16 above)
Comment (jimflores5) [reply src/chapters/databases-part2/home-logic.rst:28]: I edited the video file to cover up `NULL` and replace it with `False`.

File: src/chapters/databases-part2/home-page.rst (hunk 6 above)
Comment (jimflores5) [reply src/chapters/databases-part2/home-page.rst:52]: OK.  I think we'll leave this alone, since the students should be savvy enough by now to adjust the input field on their own.

File: src/chapters/databases-part2/game-logic.rst [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+
+Change Cells After Click
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. raw:: html
+
+   <section class="vid_box">
+      <iframe class="vid" src="https://www.youtube-nocookie.com/embed/an8sh-Xp5PQ" frameborder="1" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
+   </section>
+
```
Comment (codinglikeagirl42) [inline src/chapters/databases-part2/game-logic.rst:103]: So not a biggie, but from 1:20 to2:05 Line 59 = session.modified = True is seen.

It was moved to line 69 in the starter code

File: src/chapters/databases-part2/db-logic.rst [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+
+.. raw:: html
+
+   <section class="vid_box">
+      <iframe class="vid" src="https://www.youtube-nocookie.com/embed/q2EcgavIgAE" frameborder="1" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
+   </section>
+
+Video Summary
+-------------
+
```
Comment (codinglikeagirl42) [inline src/chapters/databases-part2/db-logic.rst:55]: Looks Great!!

File: src/chapters/databases-part2/introduction.rst [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+#. Follow the usual process to install a new virtual environment. Name the
+   new environment ``game-env``. (If you want to use a different name, 
+   that's fine. Just remember to update the ``.gitignore`` file).
+#. Activate the virtual environment and install Flask.
```
Comment (codinglikeagirl42) [inline src/chapters/databases-part2/introduction.rst:37]: If they are using GitHub, do we want to include instructions to switch the Head repository to their own?

Using:
git status
git add .
git commit -m"Initial setup"
check git branch is main, if not
git branch -m master main
git branch (check again)
git remote rm origin

Then create a new repo on GitHub and copy the ...or push an existing repository from the command line and paste into the command line.

lastly -
git push

File: src/chapters/databases-part2/home-page.rst (hunk 6 above)
Comment (codinglikeagirl42) [reply src/chapters/databases-part2/home-page.rst:52]: I figured it out, I forgot class ="menu" for my input button.

File: src/chapters/databases-part2/home-page.rst [hunk 20]
```diff
@@ -0,0 +1,106 @@
...
+
+Open your Minesweeper project and code along with the video as we build the
+HTML template:
+
+.. raw:: html
+
+   <section class="vid_box">
+      <iframe class="vid" src="https://www.youtube-nocookie.com/embed/iKkYGMHWWco" frameborder="1" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
+   </section>
+
```
Comment (codinglikeagirl42) [inline src/chapters/databases-part2/home-page.rst:73]: Again I missed an instance of Null - although you specifically state this is code given and we don't need to change anything from about 8:39 to 9:23 you are reference game_logic.py and you can see guessed = Null.  I wouldn't worry about it, just making a note, and there is some code scrolling .

File: src/chapters/databases-part2/home-logic.rst (hunk 16 above)
Comment (codinglikeagirl42) [reply src/chapters/databases-part2/home-logic.rst:28]: Looks Great!

File: src/chapters/databases-part2/game-logic.rst (hunk 17 above)
Comment (jimflores5) [reply src/chapters/databases-part2/game-logic.rst:103]: Thanks.  I see the statement, and it is also present on line 70 in the video.  These are redundant, so I probably removed the first one after making the video.  The latter is hit regardless of how the conditionals turn out.

I vote we leave this video alone.  In a follow up clip (video 7a on the Gameplay Functions page), there's a deeper explanation of the code, and this one has the offending line 59 absent.

File: src/chapters/databases-part2/db-logic.rst (hunk 18 above)
Comment (jimflores5) [reply src/chapters/databases-part2/db-logic.rst:55]: Whew!

File: src/chapters/databases-part2/home-page.rst (hunk 20 above)
Comment (jimflores5) [reply src/chapters/databases-part2/home-page.rst:73]: Thanks.  I see it.  Since the code scrolls, it would be tricky to put a patch over it.  I concur with you about leaving it alone.
